### ISSUE TITLE
Support overloaded pseudo-classes e.g. input:text and div:text(...)

### ISSUE THREAD

**Issue by chocolateboy** [fb55/css-select#78]
TL;DR: I've implemented a `:text()` filter which matches an element's text content either by string equality:

```css
span:text("IMDb")
span:text('IMDb')
```

or pattern matching (with optional flags):

```css
span:text(/^IMDb$/i)
```

I'd like to use it with Cheerio, but css-select won't let me do it without monkey-patching.

---

I'm trying to move away from XPath to CSS3/Cheerio/jQuery as much as possible, and the main thing holding me back is text selection e.g. in XPath I can easily select an element by its text content:

```xpath
//span[text()="IMDb"]
```
But it's a pain to do in Cheerio/jQuery &c:

```javascript
$('span').filter(function () { return $(this).text() === 'aaaagh!' })
```

There's also `:contains(...)` and `:icontains(...)`, but neither of them work very well on unstructured markup. Identifying nodes by their text content is usually a last resort when there are few (or no) classes or IDs to rely on, so, IMO, we need all the precision we can get.

css-select already provides a jQuery-style `:text` selector for textboxes, but for various reasons I don't want to change the name of my selector:

* it matches jQuery's `$.text()` method
* it matches XPath's `[text()]` predicate
* it matches text!

On my current project, I can happily override the built-in `:text` pseudo since I don't use it, but I'd like the filter to be available on other projects without hackery and without breaking `input:text`. Currently, adding parens to the `input:text` pseudo blows up because it's meant to be parameterless, so supporting both wouldn't break any existing code.

By exposing the filters and pseudos, css-select *almost* allows me to use both (it happily parses anything I throw at it between those parentheses :+1:), but the strict argument verification (`verifyArgs`) gets in the way i.e. what I want to do is install a callback in `filters['text']`, and inside the callback either return my function if there's a parameter, or the original function if there isn't:

```javascript
filters.text = function (next, param) {
    if (param == null) { // input:text
        let fn = pseudos.text

        if (next === boolbase.trueFunc) {
            return fn
        }

        return function text (elem) {
            return fn(elem, param) && next(elem)
        }
    } else { // span:text(...)
        let parsed = parseParam(param)

        if ({}.toString.call(parsed) === '[object RegExp]') {
            return function text (elem) {
                return next(elem) && parsed.test(getText(elem))
            }
        } else {
            return function text (elem) {
                return next(elem) && parsed === getText(elem)
            }
        }
    }
}
```

This works fine for the filter, but blows up when I try to use the pseudo:

    SyntaxError: pseudo-selector :text requires an argument

There are a couple of ways round this:

1) add the filter to css-select as a new built-in and special case it inside `verifyArgs` as is currently done for `scope`
2) allow functions to signal to `compile`/`verifyArgs` that their args shouldn't be verified
3) add a new export type alongside `filters` and `pseudos`, e.g. `plugins`, which is checked before `filters` and isn't subject to arg verification.  

2\) could be something like:

```javascript
function verifyArgs (func, name, subselect) {
    if (func.noVerifyArgs) {
        return;
    }
}
 ```
I'm happy to provide a PR for 1), but I think 2) and 3) provide more flexibility, and 2) is the smallest change and the quickest/easiest to implement.

**Comment by fb55** [comment]
This seems like a very specific issue. As the CSS spec only includes selectors with or without arguments, the current behaviour seems adequate. Is there a specific reason why you can't simply have a different name for the selector with an argument?

**Comment by chocolateboy** [comment]
> the CSS spec only includes selectors with or without arguments

True, but `:text` isn't in the CSS spec, so this isn't an issue in spec-compliant mode.

> Is there a specific reason why you can't simply have a different name for the selector with an argument?

Because naming things is hard :-)

`text` is the most obvious, most descriptive, and most commonly-used name for this. If I could think of a bearable replacement, I wouldn't have requested this. `:matches` is already used, so it would still require some sort of overloading, and doesn't specify what is being matched. `:only-text` and `:all-text` are, ambiguous. `:like` and `:match` (and `:matches`) are passable in the regex case, but read like `:contains` in the string case i.e. they suggest a substring match rather than equality. I've even considered `:txt` before coming to my senses :-)

I would argue that the current `:text` pseudo is inconsistently/confusingly named, since, unlike `:has`, `:not`, `:eq` &c. it has nothing to do with the [corresponding jQuery method](https://api.jquery.com/text/).

The only name I can think of that isn't inaccurate or barbarous is `:text-content`, and I might use that as the default for those who are happy with it, but I'd like to have the *option* to overload `:text`.

I agree that pseudo-class overloading may have few, if any, other use cases but IMO it's the optimal solution for this specific case. How about supporting the arg-verification override, but only for non-standard selectors?

```javascript
function verifyArgs (func, name, subselect) {
    if (!re_CSS3.test(name) && func.noVerifyArgs) {
        return;
    }
}
```

**Comment by chocolateboy** [comment]
Just realised I can work around this by providing a function whose length < 1 e.g. by unpacking `arguments` (or using `...args`):

```javascript
filters['text'] = function(...args) {                            
    let [ next, param ] = args
    // ...
}
```

It depends on the current implementation of `verifyArgs` ruling out invalid lengths rather than ruling in valid lengths, though, so I'd prefer to have an explicit override rather than relying on an implementation detail.
